File: learning/SparkSQL.py
from pyspark.sql import HiveContext, Row # or
from pyspark.sql import SQLContext, Row  # (if can't include hive dependencies)

sc = SparkContext(...)
hive_ctx = HiveContext(sc) # or:
sqlContext = SQLContextSingleton.getInstance(sc)


# Basic Query Example
input = hiveCtx.jsonFile(input_file)
input.registerTempTable('tweets')
top_tweets = hiveCtx.sql(
    '''
    SELECT text, retweetCount
    FROM tweets
    ORDER BY retweetCount
    LIMIT 10
    ''')


# Dataframes
# Common operations
df.show()
df.select('name', df('age') + 1)
df.filter(df('age') > 19)
df.groupBy(df('name')).min() # max, mean, agg, etc

# Convert DF to RDD
top_tweet_text = top_tweets.rdd().map(lambda row: row.text)


# From RDDs
happy_people_RDD = sc.parallelize([Row(name='REM', favorite_drink='coffee')])
happy_people_DF = hive_ctx.inferSchema(happy_people_RDD)
happy_people_DF.registerTempTable('happy_people')


# UDFs (User-Defined Functions)

File: learning/SparkSQL.scala
/* Using Spark SQL in Applications
 * Initializing Spark SQL
 */
import org.apache.spark.sql.hive.HiveContext
import org.apache.spark.sql.SQLContext // if can't have Hive dependencies

// Construct a SQL context
val conf = new SparkConf().setAppName('myApp')
val sc = new SparkContext(conf)

// Create Spark SQL Hive Context
val hiveCtx = new HiveContext(sc)
// Import implicit conversions
import hiveCtx.implicits._

// or
val sqlContext = SQLContextSingleton.getInstance(sc)


// Newer approach:
import org.apache.spark.sql.SparkSession
val spark = SparkSession.builder.getOrCreate()
import spark.implicits._



/* Basic Query Example */
val input = hiceCtx.jsonFile(inputFile)
input.registerTempTable("tweets") // register the input schema RDD
val topTweets = hiveCtx.sql(
  """
     SELECT text, retweetCount
     FROM tweets
     ORDER BY retweetCount
     LIMIT 10
  """)


/* Dataframes */
// Common operations
df.show()
df.select("name", df("age") + 1)
df.filter(df("age") > 19)
df.groupBy(df("name")).min() // max, mean, agg, etc

// convert DF to RDD
val topTweetText = topTweets.rdd().map(row => row.getString(0))


/* From RDDs */
case class HappyPerson(handle: String, favoriteDrink: String)

val happyPeopleRDD = sc.parallelize(List(HappyPerson("Michael Stipe", "boba tea")))
// implicit conversion equiv to sqlCtx.createDataFrame(happyPeopleRDD)
happyPeopleRDD.registerTempTable("happyPeople")


/* UDFs (User-Defined Funtions) */